src/fancy-transformers/fragment.ts

```diff
@@ -0,0 +1,177 @@
+    this.length=length;
+    this.id=id;
+    this.index=index;
+    this.count=count;
+    this.payload=payload;
+    this.padding=padding;
+  }
+
+  static randomId = () : ArrayBuffer => {
+    var bytes = new Uint8Array(32);
+    for (var i = 0; i < bytes.byteLength; i++) {
+      bytes[i] = Math.floor(Math.random()*255);
+    }
+    return bytes.buffer;
+  }
+
+  static decodeFragment = (buffer:ArrayBuffer) : Fragment => {
+//    log.debug('Decode fragment %1 %2', buffer.byteLength, length);
+    var parts = split_(buffer, 2);
+    var lengthBytes = parts[0];
+    var length = decodeLength_(lengthBytes);
+    buffer = parts[1];
+
+    parts=split_(buffer, 32);
+    var fragmentId=parts[0];
+    buffer=parts[1];
+
+    var fragmentNumber=takeByte_(buffer);
+    buffer=dropByte_(buffer);
+
+    var totalNumber=takeByte_(buffer);
+    buffer=dropByte_(buffer);
+
+//    log.debug('Decoded fragment %1 %2 %3', fragmentId, fragmentNumber, totalNumber);
+
+    var payload : ArrayBuffer=null;
+    var padding : ArrayBuffer=null;
+
+    if(buffer.byteLength > length) {
+      parts=split_(buffer, length);
+      payload=parts[0];
+      padding=parts[1];
+//      log.debug('shortened payoad %1 %2 %3', buffer.byteLength, length, payload.byteLength);
+    } else if (buffer.byteLength == length) {
+      payload=buffer;
+      padding=new ArrayBuffer(0);
+//      log.debug('perect payoad %1 %2 %3', buffer.byteLength, length, payload.byteLength);
+    } else { // buffer.byteLength < length
+      throw new Error("Short buffer");
+    }
+
+    var fragment=new Fragment(length, fragmentId, fragmentNumber, totalNumber, payload, padding);
+//    log.info("Decoding %1 %2 %3", fragment.id, fragment.index, fragment.count);
+    return fragment;
+  }
+
+  public encodeFragment = () : ArrayBuffer => {
+//    log.info("Encoding %1 %2 %3", this.id, this.index, this.count);
+    return this.assemble_([this.encodeLength_(this.length), this.id, encodeByte_(this.index), encodeByte_(this.count), this.payload, this.padding]);
+  }
+
+  /* Takes a number and returns a two byte (network byte order) representation
+   * of this number.
+   */
+   // TODO(bwiley): Byte order may be backward
+  private encodeLength_ = (len:number) : ArrayBuffer => {
+    var bytes = new Uint8Array(2);
+    bytes[0] = Math.floor(len >> 8);
+    bytes[1] = Math.floor((len << 8) >> 8);
+    return bytes.buffer;
+  }
+
+  private assemble_ = (buffers:ArrayBuffer[]) : ArrayBuffer => {
+    var total=0;
+    for(var i=0; i<buffers.length; i++) {
+      total=total+buffers[i].byteLength;
+    }
+
+    var result = new Uint8Array(total);
+    var toIndex=0;
+    for(var i=0; i<buffers.length; i++) {
+      var bytes=new Uint8Array(buffers[i]);
+      for(var fromIndex=0; fromIndex<buffers[i].byteLength; fromIndex++) {
+        result[toIndex]=bytes[fromIndex];
+        toIndex=toIndex+1;
```

Comment:
you can do fromIndex++, toIndex++ in the last clause of the for-loop header.  Also another good place for a general copy routine.